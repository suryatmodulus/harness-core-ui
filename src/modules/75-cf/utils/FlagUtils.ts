/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { Feature, GitDetails, PatchFeatureQueryParams, usePatchFeature } from 'services/cf'

export interface FlagPatchParams {
  accountIdentifier: string
  orgIdentifier: string
  projectIdentifier: string
  environmentIdentifier: string
}

const makeInstruction = (kind: string, variation: string, targetIdentifiers: string[], gitDetails?: GitDetails) => {
  const instructions = {
    instructions: [
      {
        kind,
        parameters: {
          variation,
          targets: targetIdentifiers
        }
      }
    ]
  }

  return gitDetails ? { ...instructions, gitDetails } : instructions
}

const makePatchHook =
  (patchKind: string) =>
  ({ accountIdentifier, orgIdentifier, projectIdentifier, environmentIdentifier }: FlagPatchParams) => {
    const { mutate } = usePatchFeature({
      identifier: '',
      queryParams: {
        account: accountIdentifier,
        accountIdentifier,
        org: orgIdentifier,
        project: projectIdentifier,
        environment: environmentIdentifier
      } as PatchFeatureQueryParams
    })

    return (featureFlag: Feature, variation: any, targetIdentifiers: string[], gitDetails?: GitDetails) => {
      const body = makeInstruction(patchKind, variation, targetIdentifiers, gitDetails)

      return mutate(body, { pathParams: { identifier: featureFlag.identifier } })
    }
  }

export const useServeFeatureFlagVariationToTargets = makePatchHook('addTargetsToVariationTargetMap')

export const useRemoveFeatureFlagVariationFromTargets = makePatchHook('removeTargetsToVariationTargetMap')
