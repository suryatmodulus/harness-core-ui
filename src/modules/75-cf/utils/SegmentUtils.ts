/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { PatchSegmentQueryParams, usePatchSegment } from 'services/cf'

export interface TargetSegmentParams {
  accountIdentifier: string
  orgIdentifier: string
  environmentIdentifier: string
  projectIdentifier: string
}

const makeInstructions = (kind: string, targetIdentifiers: string[]) => ({
  instructions: [
    {
      kind,
      parameters: {
        targets: targetIdentifiers
      }
    }
  ]
})

const makeTargetsToSegmentHook =
  (patchKind: string) =>
  ({ accountIdentifier, orgIdentifier, environmentIdentifier, projectIdentifier }: TargetSegmentParams) => {
    const { mutate } = usePatchSegment({
      identifier: '',
      queryParams: {
        account: accountIdentifier,
        accountIdentifier,
        org: orgIdentifier,
        project: projectIdentifier,
        environment: environmentIdentifier
      } as PatchSegmentQueryParams
    })

    return (segmentIdentifier: string, targetIdentifiers: string[]) => {
      const body = makeInstructions(patchKind, targetIdentifiers)
      return mutate(body, { pathParams: { identifier: segmentIdentifier } })
    }
  }

/** Add targets to an include segment list */
export const useAddTargetsToIncludeList = makeTargetsToSegmentHook('addToIncludeList')

/** Remove targets to an include segment list */
export const useRemoveTargetsFromIncludeList = makeTargetsToSegmentHook('removeFromIncludeList')

/** Add targets to an exclude segment list */
export const useAddTargetsToExcludeList = makeTargetsToSegmentHook('addToExcludeList')

/** Remove targets to an exclude segment list */
export const useRemoveTargetsFromExcludeList = makeTargetsToSegmentHook('removeFromExcludeList')
