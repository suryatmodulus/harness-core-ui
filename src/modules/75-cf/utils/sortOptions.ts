/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

export interface Option {
  label: string
  value: string
}

export const sortOptions = (options: Option[]): Option[] =>
  options.sort(({ label: a }, { label: b }) => a.localeCompare(b))
