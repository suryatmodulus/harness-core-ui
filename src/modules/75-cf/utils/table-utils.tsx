/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import type { CellProps } from 'react-table'

export function withTableData<Data extends Record<any, any>, T>(
  mapDataToProps: (data: CellProps<Data>) => T
): (Comp: React.FC<T>) => React.FC<CellProps<Data>> {
  return (Comp: React.FC<T>) => (cellProps: CellProps<Data>) => <Comp {...mapDataToProps(cellProps)} />
}
