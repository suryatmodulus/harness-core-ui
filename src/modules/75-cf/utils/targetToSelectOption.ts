import type { SelectOption } from '@wings-software/uicore'
import type { Target } from 'services/cf'

export default function targetToSelectOption({ name, identifier }: Target): SelectOption {
  return { label: name, value: identifier }
}
