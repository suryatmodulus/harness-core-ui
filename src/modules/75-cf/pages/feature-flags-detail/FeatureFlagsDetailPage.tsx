/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { ReactElement, useState } from 'react'
import { Layout, Container, PageError } from '@wings-software/uicore'
import { useParams } from 'react-router-dom'
import { GetFeatureFlagQueryParams, useGetFeatureFlag } from 'services/cf'
import { getErrorMessage } from '@cf/utils/CFUtils'
import { ContainerSpinner } from '@common/components/ContainerSpinner/ContainerSpinner'
import { useStrings } from 'framework/strings'
import { useDocumentTitle } from '@common/hooks/useDocumentTitle'
import FlagActivation from '@cf/components/FlagActivation/FlagActivation'
import FlagActivationDetails from '@cf/components/FlagActivation/FlagActivationDetails'
import useActiveEnvironment from '@cf/hooks/useActiveEnvironment'
import GitSyncActions from '@cf/components/GitSyncActions/GitSyncActions'
import { useGitSync } from '@cf/hooks/useGitSync'
import css from './FeatureFlagsDetailPage.module.scss'

const FeatureFlagsDetailPage: React.FC = () => {
  const { getString } = useStrings()
  const [skipLoading, setSkipLoading] = useState(false)
  const { orgIdentifier, projectIdentifier, featureFlagIdentifier, accountId } = useParams<Record<string, string>>()
  const { activeEnvironment } = useActiveEnvironment()

  useDocumentTitle(getString('featureFlagsText'))
  const queryParams = {
    project: projectIdentifier as string,
    environment: activeEnvironment,
    account: accountId,
    accountIdentifier: accountId,
    org: orgIdentifier
  } as GetFeatureFlagQueryParams

  const {
    data: featureFlag,
    loading,
    error,
    refetch
  } = useGetFeatureFlag({
    identifier: featureFlagIdentifier as string,
    queryParams
  })

  const gitSync = useGitSync()

  if (loading && !skipLoading) {
    return (
      <Container
        style={{
          position: 'fixed',
          top: 0,
          left: '290px',
          width: 'calc(100% - 290px)',
          height: 'calc(100% - 144px)',
          zIndex: 1
        }}
      >
        <ContainerSpinner />
      </Container>
    )
  }

  if (error) {
    return (
      <PageError
        message={getErrorMessage(error)}
        onClick={() => {
          refetch()
        }}
      />
    )
  }

  const refetchFlag = async (): Promise<void> => {
    setSkipLoading(true)
    await refetch({
      queryParams: {
        ...queryParams,
        environment: activeEnvironment
      }
    }).finally(() => {
      setSkipLoading(false)
    })
  }

  const GitSyncActionsComponent = (): ReactElement => (
    <GitSyncActions
      isLoading={gitSync.gitSyncLoading}
      branch={gitSync.gitRepoDetails?.branch || ''}
      repository={gitSync.gitRepoDetails?.repoIdentifier || ''}
      isAutoCommitEnabled={gitSync.isAutoCommitEnabled}
      isGitSyncPaused={gitSync.isGitSyncPaused}
      handleToggleAutoCommit={gitSync.handleAutoCommit}
      handleGitPause={gitSync.handleGitPause}
    />
  )

  return (
    <Container flex height="100%">
      <Layout.Vertical className={css.flagContainer} width={450}>
        {featureFlag && (
          <FlagActivationDetails
            featureFlag={featureFlag}
            refetchFlag={refetch}
            gitSyncActionsComponent={gitSync?.isGitSyncActionsEnabled ? <GitSyncActionsComponent /> : undefined}
            gitSync={gitSync}
          />
        )}
      </Layout.Vertical>

      <Layout.Horizontal
        width="calc(100% - 450px + 20px)"
        height="100%"
        style={{ transform: 'translateX(-20px)', background: 'var(--white)' }}
      >
        <Layout.Vertical width="100%">
          {!loading && featureFlag && (
            <FlagActivation
              refetchFlag={refetchFlag}
              gitSync={gitSync}
              project={projectIdentifier as string}
              flagData={featureFlag}
            />
          )}
        </Layout.Vertical>
      </Layout.Horizontal>
    </Container>
  )
}

export default FeatureFlagsDetailPage
