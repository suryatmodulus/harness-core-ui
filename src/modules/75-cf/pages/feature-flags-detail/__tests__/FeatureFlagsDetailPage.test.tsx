/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { getAllByText, getByText, render } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import mockImport from 'framework/utils/mockImport'
import FeatureFlagsDetailPage from '../FeatureFlagsDetailPage'
import mockEnvironments from '../../environments/__tests__/mockEnvironments'
import mockFeatureFlag from './mockFeatureFlag'

jest.mock('@cf/hooks/useGitSync', () => ({
  useGitSync: jest.fn(() => ({
    getGitSyncFormMeta: jest.fn().mockReturnValue({
      gitSyncInitialValues: {},
      gitSyncValidationSchema: {}
    }),
    isAutoCommitEnabled: false,
    isGitSyncEnabled: true,
    isGitSyncActionsEnabled: true,
    handleAutoCommit: jest.fn()
  }))
}))

describe('FeatureFlagsDetailPage', () => {
  test('FeatureFlagsDetailPage should render loading correctly', async () => {
    mockImport('services/cd-ng', {
      useGetEnvironmentListForProject: () => ({ loading: true, refetch: jest.fn() })
    })

    mockImport('@cf/hooks/useEnvironmentSelectV2', {
      useEnvironmentSelectV2: () => ({
        data: undefined,
        loading: true,
        error: undefined,
        refetch: jest.fn(),
        EnvironmentSelect: function EnvironmentSelect() {
          return <div />
        }
      })
    })

    mockImport('services/cf', {
      useGetFeatureFlag: () => ({
        data: undefined,
        loading: true,
        error: undefined,
        refetch: jest.fn()
      })
    })

    const { container } = render(
      <TestWrapper
        path="/account/:accountId/cf/orgs/:orgIdentifier/projects/:projectIdentifier/onboarding/detail"
        pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy' }}
      >
        <FeatureFlagsDetailPage />
      </TestWrapper>
    )

    expect(container.querySelector('[data-icon="steps-spinner"]')).toBeDefined()
    expect(getAllByText(container, /Loading, please wait\.\.\./)).toBeDefined()
  })

  test('FeatureFlagsDetailPage should render error correctly 1', async () => {
    const error = { message: 'SOME ERROR OCCURS' }
    mockImport('services/cd-ng', {
      useGetEnvironmentListForProject: () => ({ loading: true, refetch: jest.fn() })
    })

    mockImport('@cf/hooks/useEnvironmentSelectV2', {
      useEnvironmentSelectV2: () => ({
        data: undefined,
        loading: undefined,
        error,
        refetch: jest.fn(),
        EnvironmentSelect: function EnvironmentSelect() {
          return <div />
        }
      })
    })

    mockImport('services/cf', {
      useGetFeatureFlag: () => ({
        data: mockFeatureFlag,
        loading: undefined,
        error: undefined,
        refetch: jest.fn()
      })
    })

    const { container } = render(
      <TestWrapper
        path="/account/:accountId/cf/orgs/:orgIdentifier/projects/:projectIdentifier/onboarding/detail"
        pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy' }}
      >
        <FeatureFlagsDetailPage />
      </TestWrapper>
    )

    expect(getAllByText(container, error.message)).toBeDefined()
  })

  test('FeatureFlagsDetailPage should render error correctly 2', async () => {
    const error = { message: 'SOME ERROR OCCURS' }
    mockImport('services/cd-ng', {
      useGetEnvironmentListForProject: () => ({ loading: true, refetch: jest.fn() })
    })

    mockImport('@cf/hooks/useEnvironmentSelectV2', {
      useEnvironmentSelectV2: () => ({
        data: mockEnvironments.data.content,
        loading: undefined,
        error: undefined,
        refetch: jest.fn(),
        EnvironmentSelect: function EnvironmentSelect() {
          return <div />
        }
      })
    })

    mockImport('services/cf', {
      useGetFeatureFlag: () => ({
        data: undefined,
        loading: undefined,
        error,
        refetch: jest.fn()
      })
    })

    const { container } = render(
      <TestWrapper
        path="/account/:accountId/cf/orgs/:orgIdentifier/projects/:projectIdentifier/onboarding/detail"
        pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy' }}
      >
        <FeatureFlagsDetailPage />
      </TestWrapper>
    )

    expect(getAllByText(container, error.message)).toBeDefined()
  })

  test('FeatureFlagsDetailPage should render no environment', async () => {
    mockImport('services/cd-ng', {
      useGetEnvironmentListForProject: () => ({ loading: true, refetch: jest.fn() })
    })

    mockImport('@cf/hooks/useEnvironmentSelectV2', {
      useEnvironmentSelectV2: () => ({
        environments: [],
        loading: undefined,
        error: undefined,
        refetch: jest.fn(),
        EnvironmentSelect: function EnvironmentSelect() {
          return <div />
        }
      })
    })

    mockImport('services/cf', {
      useGetFeatureFlag: () => ({
        data: mockFeatureFlag,
        loading: undefined,
        error: undefined,
        refetch: jest.fn()
      })
    })

    const { container } = render(
      <TestWrapper
        path="/account/:accountId/cf/orgs/:orgIdentifier/projects/:projectIdentifier/onboarding/detail"
        pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy' }}
      >
        <FeatureFlagsDetailPage />
      </TestWrapper>
    )

    expect(getByText(container, 'cf.noEnvironment.title')).toBeDefined()
    expect(getByText(container, 'cf.noEnvironment.message')).toBeDefined()
    expect(getByText(container, 'cf.environments.create.title')).toBeDefined()
  })

  test('FeatureFlagsDetailPage should render data correctly', async () => {
    mockImport('services/cd-ng', {
      useGetEnvironmentListForProject: () => ({ loading: true, refetch: jest.fn() })
    })

    mockImport('@cf/hooks/useEnvironmentSelectV2', {
      useEnvironmentSelectV2: () => ({
        data: mockEnvironments.data.content,
        loading: undefined,
        error: undefined,
        refetch: jest.fn(),
        EnvironmentSelect: function EnvironmentSelect() {
          return <div />
        }
      })
    })

    mockImport('services/cf', {
      useGetFeatureFlag: () => ({
        data: mockFeatureFlag,
        loading: undefined,
        error: undefined,
        refetch: jest.fn()
      })
    })

    mockImport('@cf/components/FlagActivation/FlagActivation', {
      // FlagActivation is exported as `default`
      default: function FlagActivation() {
        return <div>FlagActivation</div>
      }
    })

    const { container } = render(
      <TestWrapper
        path="/account/:accountId/cf/orgs/:orgIdentifier/projects/:projectIdentifier/onboarding/detail"
        pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy' }}
      >
        <FeatureFlagsDetailPage />
      </TestWrapper>
    )

    expect(getAllByText(container, 'FlagActivation')).toBeDefined()
    expect(container).toMatchSnapshot()
  })
})
