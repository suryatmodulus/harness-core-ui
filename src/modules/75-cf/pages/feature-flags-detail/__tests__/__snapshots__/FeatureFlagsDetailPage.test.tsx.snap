// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureFlagsDetailPage FeatureFlagsDetailPage should render data correctly 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main StyledProps--flex"
    style="height: 100%;"
  >
    <div
      class="Layout--vertical StyledProps--main flagContainer"
      style="width: 450px;"
    >
      <div
        class="Layout--horizontal StyledProps--main breadcrumb"
      >
        <a
          href="/account/dummy/cf/orgs/dummy/projects/dummy/feature-flags"
          style="color: rgb(0, 146, 228); font-size: 12px;"
        >
          flag
        </a>
        <span
          style="display: inline-block;"
        >
          /
        </span>
        <span
          style="flex-grow: 1;"
        />
        <span
          class="bp3-popover-wrapper"
        >
          <span
            class="bp3-popover-target"
          >
            <button
              class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly Button--without-shadow Button--withLeftIcon"
              type="button"
            >
              <span
                class="bp3-icon StyledProps--main"
                data-icon="Options"
              >
                <svg
                  fill="none"
                  height="16"
                  viewBox="0 0 4 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M2 12.5a1 1 0 100 2 1 1 0 000-2zm-2 1a2 2 0 114 0 2 2 0 01-4 0zm2-12a1 1 0 100 2 1 1 0 000-2zm-2 1a2 2 0 114 0 2 2 0 01-4 0zM2 7a1 1 0 100 2 1 1 0 000-2zM0 8a2 2 0 114 0 2 2 0 01-4 0z"
                    fill="currentColor"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </button>
          </span>
        </span>
      </div>
      <div
        class="StyledProps--font StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-space-between StyledProps--flex-alignItems-center"
        style="width: 230px;"
      >
        <div
          class="Layout--horizontal Layout--layout-spacing-small StyledProps--main"
          style="width: 400px;"
        >
          <div
            class="StyledProps--font StyledProps--main gitRepoText"
          >
            <span
              class="bp3-icon StyledProps--main"
              data-icon="repository"
            >
              <svg
                height="16"
                viewBox="0 0 13 14"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12.438 2.452l-.006-.017a.367.367 0 00-.067-.114l-.004-.007L10.409.125a.374.374 0 00-.28-.125H2.327a.376.376 0 00-.28.125L.095 2.314l-.005.008a.374.374 0 00-.066.113l-.006.017a.37.37 0 00-.018.11v11.064c0 .206.168.374.374.374h11.707a.375.375 0 00.375-.374V2.563a.369.369 0 00-.018-.111zm-7.91 7.86h3.4v.357a.318.318 0 01-.318.318H4.846a.318.318 0 01-.318-.318v-.357zM.75 9.563V6.625h3.03v.357c0 .588.479 1.066 1.067 1.066H7.61c.588 0 1.066-.478 1.066-1.066v-.357h3.03v2.938H.75zm3.03-6.626v.357c0 .588.479 1.067 1.067 1.067H7.61c.589 0 1.067-.479 1.067-1.067v-.357h3.03v2.939H.75V2.937h3.03zm4.15 0v.357a.318.318 0 01-.318.318H4.846a.318.318 0 01-.318-.318v-.357h3.4zm-3.4 3.688h3.4v.357a.318.318 0 01-.319.317H4.846a.317.317 0 01-.318-.317v-.357zM2.493.749h7.468l1.284 1.44H1.21L2.494.748zm9.213 12.502H.75v-2.939h3.03v.357c0 .589.479 1.067 1.067 1.067H7.61c.588 0 1.066-.478 1.066-1.067v-.357h3.03v2.94z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <p
              class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--color StyledProps--color-black"
              style="--text-line-clamp: 1;"
            >
              
            </p>
          </div>
          <div
            class="StyledProps--font StyledProps--main"
          >
            <div
              class="bp3-popover-wrapper bp3-fill"
            >
              <div
                class="bp3-popover-target"
              >
                <button
                  class="branchActionButton"
                >
                  <div
                    class="StyledProps--font StyledProps--main branchActionButtonWrapper"
                  >
                    <span
                      class="bp3-icon StyledProps--main"
                      data-icon="git-new-branch"
                    >
                      <svg
                        height="15"
                        viewBox="0 0 13 16"
                        width="15"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M10.406.003a2.399 2.399 0 00-.8 4.658v.94a.8.8 0 01-.8.8H4.001a2.413 2.413 0 00-.8.14V4.66A2.399 2.399 0 002.402 0a2.399 2.399 0 00-.8 4.661v6.678a2.399 2.399 0 00.8 4.661 2.399 2.399 0 00.8-4.661v-2.54a.8.8 0 01.8-.799h4.803a2.4 2.4 0 002.402-2.4v-.939a2.399 2.399 0 00-.8-4.658z"
                          fill="currentColor"
                        />
                      </svg>
                    </span>
                    <p
                      class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey900"
                    >
                      
                    </p>
                    <div
                      class="StyledProps--font StyledProps--main gitStatusIcon"
                    >
                      <span
                        class="bp3-popover-wrapper"
                      >
                        <span
                          class="bp3-popover-target"
                        >
                          <span
                            class="bp3-icon bp3-icon-full-circle StyledProps--main autoCommitDisabled"
                            data-testid="auto-commit-status-icon"
                            icon="full-circle"
                            tabindex="0"
                          >
                            <svg
                              data-icon="full-circle"
                              height="10"
                              viewBox="0 0 16 16"
                              width="10"
                            >
                              <desc>
                                full-circle
                              </desc>
                              <path
                                d="M8 0a8 8 0 100 16A8 8 0 108 0z"
                                fill-rule="evenodd"
                              />
                            </svg>
                          </span>
                        </span>
                      </span>
                    </div>
                  </div>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="StyledProps--font StyledProps--main"
      >
        <h1
          class="StyledProps--font StyledProps--font-h1 StyledProps--main"
          style="font-weight: 600; font-size: 16px; line-height: 22px; color: rgb(28, 28, 40);"
        >
          hello-world
        </h1>
        <span
          class="StyledProps--font StyledProps--main StyledProps--inline"
          style="background-color: rgb(239, 251, 255); padding: 4px 8px; border-radius: 4px; font-size: 12px; line-height: 15px; color: rgb(34, 34, 42); border: 1px solid #d9dae6;"
        >
          idLabel
          hello_world
          <button
            style="background: none; padding: 0px; transform: translateY(-2px); cursor: pointer;"
            title="clickToCopy"
          >
            <span
              class="bp3-icon bp3-icon-duplicate StyledProps--main StyledProps--color StyledProps--color-grey350"
              icon="duplicate"
            >
              <svg
                data-icon="duplicate"
                height="12"
                viewBox="0 0 16 16"
                width="12"
              >
                <desc>
                  duplicate
                </desc>
                <path
                  d="M15 0H5c-.55 0-1 .45-1 1v2h2V2h8v7h-1v2h2c.55 0 1-.45 1-1V1c0-.55-.45-1-1-1zm-4 4H1c-.55 0-1 .45-1 1v10c0 .55.45 1 1 1h10c.55 0 1-.45 1-1V5c0-.55-.45-1-1-1zm-1 10H2V6h8v8z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </button>
        </span>
        <div
          class="Layout--vertical StyledProps--main StyledProps--margin-top-medium StyledProps--margin-bottom-xlarge"
        >
          <p
            class="StyledProps--font StyledProps--main"
            style="font-weight: 500; line-height: 14px; font-size: 10px; color: rgb(85, 87, 112); letter-spacing: 0.2px;"
          >
            cf.featureFlags.createdDate
          </p>
          <p
            class="StyledProps--font StyledProps--main"
            style="font-weight: 500; line-height: 14px; font-size: 10px; color: rgb(85, 87, 112); letter-spacing: 0.2px;"
          >
            cf.featureFlags.modifiedDate
          </p>
        </div>
        <div
          class="Layout--vertical StyledProps--main module StyledProps--padding-large StyledProps--margin-top-large"
        >
          <div
            class="Layout--horizontal StyledProps--main StyledProps--flex StyledProps--flex-align-center-center StyledProps--margin-bottom-medium"
          >
            <p
              class="StyledProps--font StyledProps--main"
              style="color: rgb(28, 28, 40); font-weight: 600; font-size: 14px; line-height: 22px;"
            >
              <span
                class="align"
                data-tooltip-id="ff_ffVariations_heading"
              >
                cf.shared.variations
              </span>
            </p>
            <span
              style="flex-grow: 1;"
            />
            <button
              class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main StyledProps--intent-primary Button--with-current-color Button--without-shadow Button--withLeftIcon"
              data-testid="open-edit-variations-modal"
              type="button"
            >
              <span
                class="bp3-icon bp3-icon-edit StyledProps--main"
                icon="edit"
              >
                <svg
                  data-icon="edit"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <desc>
                    edit
                  </desc>
                  <path
                    d="M3.25 10.26l2.47 2.47 6.69-6.69-2.46-2.48-6.7 6.7zM.99 14.99l3.86-1.39-2.46-2.44-1.4 3.83zm12.25-14c-.48 0-.92.2-1.24.51l-1.44 1.44 2.47 2.47 1.44-1.44c.32-.32.51-.75.51-1.24.01-.95-.77-1.74-1.74-1.74z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            class="Layout--vertical StyledProps--main variationsList"
          >
            <p
              class="StyledProps--font StyledProps--main StyledProps--border StyledProps--border-bottom StyledProps--border-color-grey300 StyledProps--padding-bottom-small StyledProps--flex"
              style="font-size: 14px; line-height: 20px;"
            >
              <span
                style="display: inline-block; transform: translateY(1px);"
              >
                <span
                  style="border-radius: 50%; width: 12px; height: 12px; background-color: rgb(64, 101, 160); display: inline-block;"
                />
                <span
                  style="border-radius: 50%; width: 12px; height: 12px; background-color: rgb(101, 222, 242); display: inline-block; transform: translateX(-4px); margin-right: 2px;"
                />
              </span>
              cf.boolean
              <span
                style="flex-grow: 1;"
              />
              <p
                class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey400"
              >
                cf.featureFlagDetail.variationCount
              </p>
            </p>
            <div
              class="Layout--horizontal Layout--layout-spacing-xsmall StyledProps--main variationItem"
              style="align-items: center;"
            >
              <span
                style="border-radius: 50%; width: 12px; height: 12px; background-color: rgb(64, 101, 160); display: inline-block;"
              />
              <span
                class="StyledProps--font StyledProps--main StyledProps--inline StyledProps--margin-bottom-xsmall"
                style="margin-bottom: 0px;"
              >
                True
              </span>
            </div>
            <div
              class="Layout--horizontal Layout--layout-spacing-xsmall StyledProps--main variationItem"
              style="align-items: center;"
            >
              <span
                style="border-radius: 50%; width: 12px; height: 12px; background-color: rgb(101, 222, 242); display: inline-block;"
              />
              <span
                class="StyledProps--font StyledProps--main StyledProps--inline StyledProps--margin-bottom-xsmall"
                style="margin-bottom: 0px;"
              >
                False
              </span>
            </div>
          </div>
        </div>
        <div
          class="StyledProps--font StyledProps--main collapseFeatures module"
        >
          <div
            class="Collapse--main"
          >
            <div
              class="StyledProps--font StyledProps--main CollapseHeader--main Collapse--header"
            >
              <div
                class="StyledProps--font StyledProps--main CollapseHeader--leftSection"
              >
                <span
                  class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-grey400"
                  data-icon="main-chevron-right"
                >
                  <svg
                    height="10"
                    viewBox="0 0 8 14"
                    width="10"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M7.7 6.3l-6-6C1.5.1 1.3 0 1 0 .4 0 0 .4 0 1c0 .3.1.5.3.7L5.6 7 .3 12.3c-.2.2-.3.4-.3.7 0 .6.4 1 1 1 .3 0 .5-.1.7-.3l6-6c.2-.2.3-.4.3-.7s-.1-.5-.3-.7z"
                      fill="#4d5969"
                    />
                  </svg>
                </span>
                <p
                  class="StyledProps--font StyledProps--main"
                  style="font-size: 14px; color: rgb(34, 34, 42); font-weight: 600; line-height: 20px;"
                >
                  <span
                    class="align"
                    data-tooltip-id="ff_ffPrerequisites_heading"
                  >
                    cf.shared.prerequisites
                  </span>
                  <span
                    style="font-size: 12px; font-weight: 400; display: inline-block;"
                  >
                    cf.featureFlags.prerequisitesDesc
                  </span>
                </p>
              </div>
              <div
                class="StyledProps--font StyledProps--main"
              />
            </div>
            <div
              class="Collapse--collapse"
            >
              <div
                class="bp3-collapse collapse"
              >
                <div
                  aria-hidden="false"
                  class="bp3-collapse-body"
                  style="transform: translateY(-undefinedpx);"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="Layout--horizontal StyledProps--main"
      style="transform: translateX(-20px); width: calc(100% - 450px + 20px); height: 100%;"
    >
      <div
        class="Layout--vertical StyledProps--main"
        style="width: 100%;"
      >
        <div>
          FlagActivation
        </div>
      </div>
    </div>
  </div>
</div>
`;
